arrabiata/src/interpreter.rs

```diff
@@ -0,0 +1,962 @@
+//! columns.
+//! At the moment, all constraints must be of maximum degree
+//! [crate::MAX_DEGREE], but it might change in the future.
+//!
+//! The implementation relies on a representation of the circuit as a 2D array
+//! of "data points" the interpreter can use.
+//!
+//! An interpreter defines what a "position" is in the circuit and allow to
+//! perform operations using these positions.
+//! Some of these positions will be considered as public inputs and might be
+//! fixed at setup time while making a proof, when other will be considered as
+//! private inputs.
+//!
+//! On top of these abstraction, gadgets are implemented.
+//! For the Nova-like IVC schemes, we describe below the different gadgets and
+//! how they are implemented with this abstraction.
+//!
+//! **Table of contents**:
+//! - [Gadgets implemented](#gadgets-implemented)
+//!   - [Elliptic curve addition](#elliptic-curve-addition)
+//!     - [Gadget layout](#gadget-layout)
+//!   - [Hash - Poseidon](#hash---poseidon)
+//!     - [Gadget layout](#gadget-layout-1)
+//!   - [Elliptic curve scalar multiplication](#elliptic-curve-scalar-multiplication)
+//!     - [Gadget layout](#gadget-layout-2)
+//!   - [Bit composition instruction](#bit-composition-instruction)
+//! - [Handle the combinaison of constraints](#handle-the-combinaison-of-constraints)
+//! - [Permutation argument](#permutation-argument)
+//! - [Fiat-Shamir challenges](#fiat-shamir-challenges)
+//!
+//! ## Gadgets implemented
+//!
+//! ### Elliptic curve addition
+//!
+//! The Nova augmented circuit requires to perform elliptic curve operations, in
+//! particular additions and scalar multiplications.
+//!
+//! To reduce the number of operations, we consider the affine coordinates.
+//! As a reminder, here the equations to compute the addition of two different
+//! points `P1 = (X1, Y1)` and `P2 = (X2, Y2)`. Let define `P3 = (X3, Y3) = P1 +
+//! P2`.
+//!
+//! ```text
+//! - λ = (Y1 - Y2) / (X1 - X2)
+//! - X3 = λ^2 - X1 - X2
+//! - Y3 = λ (X1 - X3) - Y1
+//! ```
+//!
+//! Therefore, the addition of elliptic curve points can be computed using the
+//! following degree-2 constraints
+//!
+//! ```text
+//! - Constraint 1: λ (X1 - X2) - Y1 + Y2 = 0
+//! - Constraint 2: X3 + X1 + X2 - λ^2 = 0
+//! - Constraint 3: Y3 - λ (X1 - X3) + Y1 = 0
+//! ```
+//!
+//! If the points are the same, the λ is computed as follows:
+//!
+//! ```text
+//! - λ = (3 X1^2 + a) / (2Y1)
+//! ```
+//!
+//! #### Gadget layout
+//!
+//! For given inputs (x1, y1) and (x2, y2), the layout will be as follow:
+//!
+//! ```text
+//! | C1 | C2 | C3 | C4 | C5 | C6 | C7 | C8 | C9 | C10 | C11 | C12 | C13 | C14 | C15 | C16 | C17 |
+//! | -- | -- | -- | -- | -- | -- | -- | -- | -- | --- | --- | --- | --- | --- | --- | --- | --- |
+//! | x1 | y1 | x2 | y2 | b0 | λ  | x3 | y3 |    |     |     |     |     |     |     |     |     |
+//! ```
+//!
+//! where `b0` is equal two `1` if the points are the same, and `0` otherwise.
```

Comment:
I would separate the double and the add, it will give a much better scalar mul
And most of time we have an overwhelming chance of not doubling 'by chance', eg. when adding two commitments to different polynomial.